modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,144 @@
+package org.tessellation.dag.l0.domain.snapshot.programs
+
+import cats.MonadThrow
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.option._
+import cats.syntax.traverse._
+import cats.syntax.unorderedFoldable._
+
+import org.tessellation.dag.l0.infrastructure.snapshot.{DAGEvent, GlobalSnapshotEvent, StateChannelEvent}
+import org.tessellation.ext.kryo.RefinedSerializer
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.node.shared.domain.event.EventCutter
+import org.tessellation.schema.Block
+import org.tessellation.schema.gossip.Ordinal
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.security.hash._
+import org.tessellation.security.{Hashed, Hasher}
+
+import eu.timepit.refined.types.numeric.PosInt
+
+object GlobalSnapshotEventCutter {
+  type GlobalPeerIdEventMap = Map[PeerId, List[(Ordinal, GlobalSnapshotEvent)]]
+
+  def make[F[_]: Async: KryoSerializer: Hasher](maxBinarySizeBytes: PosInt): EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] =
+    new EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] {
+
+      def cut(events: GlobalPeerIdEventMap): F[GlobalPeerIdEventMap] =
+        needsCutting(events)
+          .ifM(
+            OrderedEvents(events)
+              .map(Iterator.unfold(_)(_.next))
+              .flatMap(removeEvents(events, _)),
+            events.pure[F]
+          )
+
+      def needsCutting(events: GlobalPeerIdEventMap): F[Boolean] =
+        events.toBinary.liftTo[F].map(_.length > maxBinarySizeBytes.value)
+
+      def removeEvents(startingEvents: GlobalPeerIdEventMap, iterator: Iterator[PeerIdEvent]): F[GlobalPeerIdEventMap] =
+        startingEvents.tailRecM { events =>
+          for {
+            eventToRemove <- MonadThrow[F].fromOption(iterator.nextOption(), new IllegalStateException("No GlobalSnapshotEvent events"))
+            reducedEvents = removeOneEvent(events, eventToRemove)
+            cutAgain <- needsCutting(reducedEvents)
+          } yield Either.cond(!cutAgain, reducedEvents, reducedEvents)
+        }
+
+      def removeOneEvent(events: GlobalPeerIdEventMap, eventToRemove: PeerIdEvent): GlobalPeerIdEventMap =
+        events.updatedWith(eventToRemove.id) {
+          case None => None
+          case Some(globalPeerIdEvent) =>
+            val globalSnapshotEvent = eventToRemove.toGlobalSnapshotEvent
+            val updatedEvents = globalPeerIdEvent.filterNot { case (_, gse) => gse == globalSnapshotEvent }
+            Option.when(updatedEvents.nonEmpty)(updatedEvents)
+        }
+
+      trait PeerIdEvent {
+        def id: PeerId
+        def fee: Long
+        def hash: Hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent
+      }
+
+      val ordering: Ordering[PeerIdEvent] = Ordering.by[PeerIdEvent, (Long, Hash)](e => (e.fee, e.hash))
```

Comment:
Ordering by `fee` and `hash` only is faulty. You can have a newer event with higher fee be attempted before the older event that actually needs to be processed first. Hash can only be used to provide deterministicity for events with the same ordering in regards to all other parameters. Blocks don't have a total order per se. Metagraph snapshots do have and ordering in regards to the chain of snapshots, but you need to discover that chain to know the order. Hashes themselves don't preserve the original order at all, they are only a representation of data underneath.

Reply:
@ryle-goehausen will need to answer this; my understanding is that ordering by hash is to make cutting deterministic in the event of a tie with the other rules

Reply:
Yes ordering by hash as a deterministic measure is fine, but the only other parameter we sort by here is the `fee`. These events aren't ordered after pull especially that there is no total order for blocks.

Reply:
> You can have a newer event with higher fee be attempted before the older event that actually needs to be processed first. 

This is as desired, especially for dynamic fee calculations, as it follows an RBF (replace by fee) policy, which we'd want to achieve for metagraph snapshot submissions. Since we're dealing with 'unordered' data in the sense that it hasn't yet been accepted, there is no 'newer' event from the perspective of the chain data.

> Hash can only be used to provide deterministicity for events with the same ordering in regards to all other parameters.

Agreed, it should always be the last sort parameter.

>Blocks don't have a total order per se. Metagraph snapshots do have and ordering in regards to the chain of snapshots, but you need to discover that chain to know the order. 

For this, we need to rely on validation to reject invalid ordering. So it should be considered invalid by ordinal definitions, NOT by mempool event receipt order. 

Then if we had the chain of events of considering a mg snapshot of ordinal 100 with fee 1 and another with ordinal 100 and fee 2, even if the second event was 'produced later' (in test time,) it should replace the first event, if all other validations are equal.

Reply:
Yeah replacing one event by another concurring with higher fee is ok and I think we should have such a mechanism. The problem is that it's not the case I'm describing here. Ordering events by fee and hash only inflicts that we can treat a snapshot with ordinal 100 and fee 10 as the first to accept over snapshot 99 with fee 5 and it won't work. For the concurring events so two or more snapshots at ordinal 100 for the same metagraph we should take the one with higher fee I agree with that.

Essentially we need to first find the order in which all the events we want to process need to be processed (which are parallel and which rely on each other) and then we can pick the ones with higher fee first (but we need to respect that primary order of events).

Reply:
>Essentially we need to first find the order in which all the events we want to process need to be processed (which are parallel and which rely on each other) and then we can pick the ones with higher fee first (but we need to respect that primary order of events).

I think the way we impose this order is through the rejection of events considered to be invalid. I.e. the future submission of events, we can discuss how to deal with this in more detail, because I realize it's the same case as a Batch DAGTransaction submission (i.e. chaining ordinals together,) which makes it more complicated. Let's talk about this in the meeting, because as far as I can tell, it's the same type of problem for which we need to rely on either some batch mechanism, or purely validation / sync operation.